azkaban-web-server/src/main/java/azkaban/flowtrigger/TriggerInstanceProcessor.java

```diff
@@ -88,22 +91,58 @@ private void executeFlowAndUpdateExecID(final TriggerInstance triggerInst) {
   }
 
   private String generateFailureEmailSubject(final TriggerInstance triggerInstance) {
-    final String flowFullName =
-        triggerInstance.getProjectName() + "." + triggerInstance.getFlowId();
-    return String.format(FAILURE_EMAIL_SUBJECT, flowFullName, this.emailer.getAzkabanName());
+    return String.format(FAILURE_EMAIL_SUBJECT, triggerInstance.getFlowId(), triggerInstance
+        .getProjectName(), this.emailer.getAzkabanName());
   }
 
-  private String generateFailureEmailBody(final TriggerInstance triggerInstance) {
-    final String triggerInstFullName =
-        triggerInstance.getProjectName() + "." + triggerInstance.getFlowId();
-    return String.format(FAILURE_EMAIL_BODY, triggerInstFullName);
+  private EmailMessage createFlowTriggerEmailMessage(final TriggerInstance triggerInst) {
+    final EmailMessage message = this.emailer.createEmailMessage(generateFailureEmailSubject
+        (triggerInst), "text/html", triggerInst.getFailureEmails());
+    final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+
+    message.addAllToAddress(triggerInst.getFailureEmails());
+    message.setMimeType("text/html");
+    message.println("<table>");
+    message.println("<tr><td>Start Time</td><td>");
+    message.println("<tr><td>" + sdf.format(new Date(triggerInst.getStartTime())) + "</td><td>");
+
+    message.println("<tr><td>End Time</td><td>");
+    message.println("<tr><td>" + sdf.format(new Date(triggerInst.getEndTime())) + "</td><td>");
+    message.println("<tr><td>Duration</td><td>"
+        + Utils.formatDuration(triggerInst.getStartTime(), triggerInst.getEndTime())
+        + "</td></tr>");
+    message.println("<tr><td>Status</td><td>" + triggerInst.getStatus() + "</td></tr>");
+    message.println("</table>");
+    message.println("");
+    final String executionUrl =
+        this.emailer.getScheme() + "://" + this.emailer.getClientHostname() + ":" +
+            this.emailer.getClientPortNumber() + "/" + "executor?" + "triggerinstanceid="
```

Comment:
Would it be better to refactor this to `this.emailer.getAzkabanUrl() + "/executor?triggerinstanceid="`? `getAzkabanUrl()` could then be used also in Emailer itself, and wouldn't need to expose all those fields.

Reply:
good point.